Mathlib/Topology/ContinuousFunction/FunctionalCalculus.lean

```diff
@@ -670,3 +683,101 @@ lemma cfc_comp_neg (hf : ContinuousOn f ((- ·) '' (spectrum R (a : A))) := by c
+
+lemma cfc_mono {f g : R → R} {a : A} (h : ∀ x ∈ spectrum R a, f x ≤ g x)
+    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac)
+    (hg : ContinuousOn g (spectrum R a) := by cfc_cont_tac) :
+    cfc f a ≤ cfc g a := by
+  by_cases ha : p a
+  · rw [cfc_apply f a, cfc_apply g a]
+    exact cfcHom_mono ha fun x ↦ h x.1 x.2
+  · simp only [cfc_apply_of_not_predicate _ ha, le_rfl]
+
+lemma cfc_nonneg_iff (f : R → R) (a : A) (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac)
+    (ha : p a := by cfc_tac) : 0 ≤ cfc f a ↔ ∀ x ∈ spectrum R a, 0 ≤ f x := by
+  rw [cfc_apply .., cfcHom_nonneg_iff, ContinuousMap.le_def]
+  simp
+
+lemma cfc_nonneg {f : R → R} {a : A} (h : ∀ x ∈ spectrum R a, 0 ≤ f x) :
+    0 ≤ cfc f a := by
+  by_cases hf : ContinuousOn f (spectrum R a)
+  · simpa using cfc_mono h
+  · simp only [cfc_apply_of_not_continuousOn _ hf, le_rfl]
+
+lemma cfc_nonpos (f : R → R) (a : A) (h : ∀ x ∈ spectrum R a, f x ≤ 0) :
+    cfc f a ≤ 0 := by
+  by_cases hf : ContinuousOn f (spectrum R a)
+  · simpa using cfc_mono h
+  · simp only [cfc_apply_of_not_continuousOn _ hf, le_rfl]
+
+lemma cfc_le_algebraMap (f : R → R) (r : R) (a : A) (h : ∀ x ∈ spectrum R a, f x ≤ r)
+    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac) (ha : p a := by cfc_tac) :
+    cfc f a ≤ algebraMap R A r :=
+  cfc_const r a ▸ cfc_mono h
+
+lemma algebraMap_le_cfc (f : R → R) (r : R) (a : A) (h : ∀ x ∈ spectrum R a, r ≤ f x)
+    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac) (ha : p a := by cfc_tac) :
+    algebraMap R A r ≤ cfc f a :=
+  cfc_const r a ▸ cfc_mono h
+
+lemma cfc_le_one (f : R → R) (a : A) (h : ∀ x ∈ spectrum R a, f x ≤ 1) : cfc f a ≤ 1 := by
+  apply cfc_cases (· ≤ 1) _ _ (by simpa using star_mul_self_nonneg (1 : A)) fun hf ha ↦ ?_
+  rw [← map_one (cfcHom ha (R := R))]
+  apply cfcHom_mono ha
+  simpa [ContinuousMap.le_def] using h
+
+lemma one_le_cfc (f : R → R) (a : A) (h : ∀ x ∈ spectrum R a, 1 ≤ f x)
+    (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac) (ha : p a := by cfc_tac) :
+    1 ≤ cfc f a := by
+  simpa using algebraMap_le_cfc f 1 a h
+
+end Semiring
+
+section Ring
+
+variable {R A : Type*} {p : A → Prop} [OrderedCommRing R] [StarRing R]
+variable [MetricSpace R] [TopologicalRing R] [ContinuousStar R]
+variable [∀ (α) [TopologicalSpace α], StarOrderedRing C(α, R)]
+variable [TopologicalSpace A] [Ring A] [StarRing A] [PartialOrder A] [StarOrderedRing A]
+variable [Algebra R A] [StarModule R A] [ContinuousFunctionalCalculus R p]
+variable [NonnegSpectrumClass R A]
+
+lemma cfcHom_le_iff {a : A} (ha : p a) {f g : C(spectrum R a, R)} :
+    cfcHom ha f ≤ cfcHom ha g ↔ f ≤ g := by
+  rw [← sub_nonneg, ← map_sub, cfcHom_nonneg_iff, sub_nonneg]
+
+lemma cfc_le_iff (f g : R → R) (a : A) (hf : ContinuousOn f (spectrum R a) := by cfc_cont_tac)
+    (hg : ContinuousOn g (spectrum R a) := by cfc_cont_tac) (ha : p a := by cfc_tac) :
+    cfc f a ≤ cfc g a ↔ ∀ x ∈ spectrum R a, f x ≤ g x := by
```

Comment:
Are these two lemmas false for `ℝ≥0`...?

Reply:
No, but unfortunately you have to go through \R to prove them. That's because when we have a generic semiring we don't have a subtraction. In NNReal we do, and things can play nice, but we can't encapsulate that with existing typeclasses.

Reply:
Alright, I see.